src/pybind/mgr/cephadm/exchange.py

```diff
@@ -0,0 +1,164 @@
+    deployed_by = _DataField(field_type=_or_none(list))
+    rank = _DataField(field_type=_or_none(int))
+    rank_generation = _DataField(field_type=_or_none(int))
+    extra_container_args = _DataField(field_type=_or_none(list))
+    extra_entrypoint_args = _DataField(field_type=_or_none(list))
+
+    def __init__(
+        self,
+        init_data: Optional[Dict[str, Any]] = None,
+        *,
+        service_name: str = "",
+        ports: Optional[List[int]] = None,
+        ip: Optional[str] = None,
+        deployed_by: Optional[List[str]] = None,
+        rank: Optional[int] = None,
+        rank_generation: Optional[int] = None,
+        extra_container_args: Optional[List[str]] = None,
+        extra_entrypoint_args: Optional[List[str]] = None,
+    ):
+        self.data = dict(init_data or {})
+        # set fields
+        self.service_name = service_name
+        self.ports = ports or []
+        self.ip = ip
+        self.deployed_by = deployed_by
+        self.rank = rank
+        self.rank_generation = rank_generation
+        self.extra_container_args = extra_container_args
+        self.extra_entrypoint_args = extra_entrypoint_args
+
+    def get_data(self) -> Dict[str, Any]:
+        return self.data
+
+    to_simplified = get_data
+
+    @classmethod
+    def convert(
+        cls,
+        value: Union[Dict[str, Any], "DeployMeta", None],
+    ) -> "DeployMeta":
+        if not isinstance(value, DeployMeta):
+            return cls(value)
+        return value
+
+
+class Deploy:
+    """Set of fields that instructs cephadm to deploy a
+    service/daemon.
+    """
+
+    fsid = _DataField(field_type=str)
+    name = _DataField(field_type=str)
+    image = _DataField(field_type=str)
+    deploy_arguments = _DataField(field_type=list)
+    params = _DataField(field_type=dict)
+    meta = _DataField(field_type=DeployMeta.convert)
+    config_blobs = _DataField(field_type=dict)
+
+    def __init__(
+        self,
+        init_data: Optional[Dict[str, Any]] = None,
+        *,
+        fsid: str = "",
+        name: str = "",
+        image: str = "",
+        deploy_arguments: Optional[List[str]] = None,
+        params: Optional[Dict[str, Any]] = None,
+        meta: Optional[DeployMeta] = None,
+        config_blobs: Optional[Dict[str, Any]] = None,
+    ):
+        self.data = dict(init_data or {})
+        # set fields
+        self.fsid = fsid
+        self.name = name
+        self.image = image
+        self.deploy_arguments = deploy_arguments or []
+        self.params = params or {}
+        self.meta = DeployMeta.convert(meta)
+        self.config_blobs = config_blobs or {}
+
+    def get_data(self) -> Dict[str, Any]:
+        """Return the underlying data dict."""
+        return self.data
+
+    def to_simplified(self) -> Dict[str, Any]:
+        """Return a simplified serializable version of the object."""
+        return {k: _get_data(v) for k, v in self.get_data().items()}
```

Comment:
I don't fully understand how attributes other than `data` get accounted for here. If everything is in the `data` attribute dict  it makes sense it would get that dict, but if I just do
```
x = Deploy(fsid='aaa')
x.dump_json_str()
```
how is that `fsid` attribute getting into the resulting json string?

Reply:
https://github.com/ceph/ceph/pull/51738/files#diff-d3250effa3c68ed2ca9a9a8d471d640082fb191dfbfc3dc0500a0aa936fb3d9bR146

When you pass one of these keyword-only arguments to __init__  we then assign to `self.fsid` in the init function. This assignment is to an existing descriptor. The descriptors `__get__` method is called and thus the value is stored in the data dict rather than as a typical value attr of the class.

Reply:
ah, okay. So `self.fsid = fsid` in the `Deploy` class actually goes through `__set__` in the `_DataField` class which is really setting `self.data['fsid']` in the `Deploy` class instance? I guess that implies `obj` param in the `__set__` function for `_DataField` is picking up the class the `_DataField` is an attribute of somehow? I think that's where I got lost.

Reply:
okay, after reading the docs on descriptors this makes more sense